include/pisa/query/algorithm/maxscore_query.hpp

```diff
@@ -1,83 +1,140 @@
+                --first_upper_bound;
+                if (first_lookup == cursors.begin()) {
+                    return UpdateResult::ShortCircuit;
                 }
             }
+            return UpdateResult::Continue;
+        };
+
+        if (update_non_essential_lists() == UpdateResult::ShortCircuit) {
+            return;
+        }
 
-            // try to complete evaluation with non-essential lists
-            for (size_t i = non_essential_lists - 1; i + 1 > 0; --i) {
-                if (!m_topk.would_enter(score + upper_bounds[i])) {
-                    break;
+        float current_score = 0;
+        std::uint32_t current_docid = 0;
+
+        while (current_docid < max_docid) {
+            auto status = DocumentStatus::Skip;
+            while (status == DocumentStatus::Skip) {
+                if (PISA_UNLIKELY(next_docid >= max_docid)) {
+                    return;
                 }
-                ordered_cursors[i]->next_geq(cur_doc);
-                if (ordered_cursors[i]->docid() == cur_doc) {
-                    score += ordered_cursors[i]->score();
+
+                current_score = 0;
+                current_docid = std::exchange(next_docid, max_docid);
+
+                std::for_each(cursors.begin(), first_lookup, [&](auto& cursor) {
+                    if (cursor.docid() == current_docid) {
+                        current_score += cursor.score();
+                        cursor.next();
+                    }
+                    if (auto docid = cursor.docid(); docid < next_docid) {
+                        next_docid = docid;
+                    }
+                });
+
+                status = DocumentStatus::Insert;
+                auto lookup_bound = first_upper_bound;
+                for (auto pos = first_lookup; pos != cursors.end(); ++pos, ++lookup_bound) {
+                    auto& cursor = *pos;
+                    if (not above_threshold(current_score + *lookup_bound)) {
+                        status = DocumentStatus::Skip;
+                        break;
+                    }
+                    cursor.next_geq(current_docid);
+                    if (cursor.docid() == current_docid) {
+                        current_score += cursor.score();
+                    }
                 }
             }
-
-            if (m_topk.insert(score, cur_doc)) {
-                update_non_essential_lists();
+            if (m_topk.insert(current_score, current_docid)
+                && update_non_essential_lists() == UpdateResult::ShortCircuit) {
+                return;
             }
+        }
+    }
 
-            cur_doc = next_doc;
+    template <typename Cursors>
+    void operator()(Cursors&& cursors_, uint64_t max_docid)
+    {
+        using cursor_type = typename std::decay_t<Cursors>::value_type;
+        if (cursors_.empty()) {
+            return;
         }
+        auto cursors = sorted(cursors_);
+        run_sorted(cursors, max_docid);
+        std::swap(cursors, cursors_);
```

Comment:
Yes, it will prevent passing `const&` but we can't pass const anyway because we need to use the cursors. For performance reasons I mentioned before, I move out the cursors of the original container. Because I do that, I'm giving it back at the end. Otherwise range queries won't work.